### PR TITLE
Define file and position based deletion file in spec

### PR DESCRIPTION
**Pull request by chenjunjiedada**
This is for https://github.com/apache/incubator-iceberg/issues/359

### CODE REVIEW

File: site/docs/spec.md [hunk 1]
```diff
@@ -232,6 +232,7 @@ The schema of a manifest file is a struct called `manifest_entry` with the follo
 | **`128  upper_bounds`**           | `optional map<129: int, 130: binary>` | Map from column id to upper bound in the column serialized as binary [1]. Each value must be greater than or equal to all values in the column for the file.                                         |
 | **`131  key_metadata`**           | `optional binary`                     | Implementation-specific key metadata for encryption                                                                                                                                                  |
 | **`132  split_offsets`**          | `optional list`                       | Split offsets for the data file. For example, all row group offsets in a Parquet file. Must be sorted ascending.                                                                                     |
+| **`134  file_type`**              | `optional int`                        | Type of the data file. `0`: normal data file that has the same schema as the table's schema. `1`: file and position based deletion file.                                                         |
```
Comment (rdblue) [inline site/docs/spec.md:235]: This change should not be in this PR. The purpose of this PR is to define the format of a file and position delete file, not to make additional changes to metadata. Could you remove this?

File: site/docs/spec.md [hunk 2]
```diff
@@ -254,6 +255,33 @@ Notes:
...
+#### Deletion Files
+
+Deletion files are files that indicate deletions of pre-existing rows to be applied to the dataset at read time. Deletion files may either specify rows by column value or by file name and row position.
+
+1. The file and position based deletion file has the schema as following:
+```
+{
+  filename string,
+  position long
+}
```
Comment (rdblue) [inline site/docs/spec.md:267]: Could you update this to be like the other schemas defined in the spec? It should be a table with a description of each field, Iceberg types, and should include field IDs.

File: site/docs/spec.md [hunk 3]
```diff
@@ -254,6 +255,33 @@ Notes:
...
+file1, 5
+file2, 3
+file2, 4
+file2, 7
+file3, 6
+file3, 8
+file3, 9
+```
+ 
+It is also worth to note that in order to keep module independence, deletion files are written with the same file format as the table's file format.
```
Comment (rdblue) [inline site/docs/spec.md:283]: This is a recommendation not a requirement, so we should specifically say that. The requirement is that a delete file can be written using any supported data file format.

Also, the purpose of the recommendation is not module independence. People choose file formats based on what they use for most tables and have experience tuning, so it makes sense to use the same format for delete files and delta files.

It's convenient to not need to build a service with a dependency on iceberg-orc or iceberg-parquet if all data and delete files are Avro, but we don't want to have people misinterpret the spec and think that there is a guarantee that delete file formats match data files.

File: site/docs/spec.md (hunk 2 above)
Comment (rdblue) [reply site/docs/spec.md:267]: Also, these fields are `required`.

File: site/docs/spec.md [hunk 4]
```diff
@@ -254,6 +255,33 @@ Notes:
...
+file1, 1
+file1, 2
+file1, 5
+file2, 3
+file2, 4
+file2, 7
+file3, 6
+file3, 8
+file3, 9
+```
```
Comment (rdblue) [inline site/docs/spec.md:281]: I don't think this is necessary. It should be sufficient to say that the delete file must be sorted by ascending filename then position.

This is for two reasons:
1. Sorting by file allows filter pushdown by file in columnar storage formats.
2. Sorting by position allows filtering rows while scanning, to avoid keeping deletes in memory.

I think it would help to rephrase "so as to leverage the merge sort" to "allow filtering rows while scanning" or something similar. Although we think of this as merge-sort, the operation is not a sort that produces a sorted list of deletes and rows -- it's a filter.

File: site/docs/spec.md [hunk 5]
```diff
@@ -254,6 +255,33 @@ Notes:
 
 1. Technically, data files can be deleted when the last snapshot that contains the file as “live” data is garbage collected. But this is harder to detect and requires finding the diff of multiple snapshots. It is easier to track what files are deleted in a snapshot and delete them when that snapshot expires.
 
+#### Deletion Files
```
Comment (rdblue) [inline site/docs/spec.md:258]: This section should be "Position-based delete files" to distinguish these from equality delete files that will be added later.

File: site/docs/spec.md [hunk 6]
```diff
@@ -254,6 +255,33 @@ Notes:
...
+
+Deletion files are files that indicate deletions of pre-existing rows to be applied to the dataset at read time. Deletion files may either specify rows by column value or by file name and row position.
```
Comment (rdblue) [inline site/docs/spec.md:260]: Let's make this specific to position-based delete files instead of describing delete files generally. How about this:

> Position-based delete files identify rows in one or more data files that have been deleted.

File: site/docs/spec.md [hunk 7]
```diff
@@ -254,6 +255,33 @@ Notes:
...
+
+1. The file and position based deletion file has the schema as following:
```
Comment (rdblue) [inline site/docs/spec.md:262]: This needs to clearly define what file and position are. I think that `file` should be renamed to `file_path` to match tracking in the manifest file and should use a similar description, along with a note that it must match what's in the manifest. We also need to note that positions start at 0.

* `file_path` - The full URI of a data file, with FS scheme. This must match the `file_path` of the target data file in a manifest entry.
* `position` - The ordinal position of a deleted row in the target data file identified by `file_path`, starting at 0.

**Comment by rdblue** [comment]
Thanks, @chenjunjiedada! This looks like a good start. I've added a few comments where I'd like to make the spec more clear.

File: site/docs/spec.md (hunk 5 above)
Comment (rdblue) [inline site/docs/spec.md:258]: Why is this a sub-section of _Manifests_?

Let's create a new section after _Table Metadata_ called _Delete Format_, with a sub-section for _Position-based delete files_. Then we should note that content in that section is in preparation for incrementing the format version:

> This section details how to encode row-level deletes in Iceberg metadata. Row-level deletes are not supported in the current format version, 1. This part of the spec is not yet complete and will be completed as format version 2.

File: site/docs/spec.md (hunk 1 above)
Comment (chenjunjiedada) [reply site/docs/spec.md:235]: Sure, will move this to the file_type PR.

File: site/docs/spec.md (hunk 2 above)
Comment (chenjunjiedada) [reply site/docs/spec.md:267]: Sure, no problem.

File: site/docs/spec.md (hunk 5 above)
Comment (chenjunjiedada) [reply site/docs/spec.md:258]: Make sense!

File: site/docs/spec.md (hunk 3 above)
Comment (chenjunjiedada) [reply site/docs/spec.md:283]: I see, let me rephrase this.

File: site/docs/spec.md (hunk 6 above)
Comment (chenjunjiedada) [reply site/docs/spec.md:260]: OK, let me update this.

File: site/docs/spec.md (hunk 2 above)
Comment (chenjunjiedada) [reply site/docs/spec.md:267]: You mean json format, right?

File: site/docs/spec.md (hunk 7 above)
Comment (chenjunjiedada) [reply site/docs/spec.md:262]: I put the description into `doc` property of the field.

File: site/docs/spec.md (hunk 5 above)
Comment (chenjunjiedada) [reply site/docs/spec.md:258]: no problem.

File: site/docs/spec.md (hunk 4 above)
Comment (chenjunjiedada) [reply site/docs/spec.md]: Ok, let me update this.

File: site/docs/spec.md (hunk 2 above)
Comment (rdblue) [reply site/docs/spec.md]: No, I meant table. Look at the other struct type definitions and copy those. The table format makes it easy to read.

File: site/docs/spec.md [hunk 8]
```diff
@@ -333,6 +333,38 @@ Table metadata is stored as JSON. Each table metadata change creates a new table
...
+  } ]
+}
+```
+
+The rows in the delete file must be sorted by `file_path` then `position` to optimize filtering rows while scanning. 
+
+*  Sorting by `file_path` allows filter pushdown by file in columnar storage formats.
+*  Sorting by `position` allows filtering rows while scanning, to avoid keeping deletes in memory.
+ 
+Though the delete files can be written using any supported data file format in Iceberg, it is recommended to write delete files with same file format as the table's file format to keep module independence.
```
Comment (rdblue) [inline site/docs/spec.md:367]: Let's remove "to keep module independence" because it isn't clear what that means.

File: site/docs/spec.md [hunk 9]
```diff
@@ -333,6 +333,38 @@ Table metadata is stored as JSON. Each table metadata change creates a new table
...
+Position-based delete files identify rows in one or more data files that have been deleted. It has the schema as following:
+```json
+{
+  "type": "struct",
+  "fields": [ {
+    "id": 1,
+    "name": "file_path",
+    "required": true,
+    "type": "string",
+    "doc": "The full URI of a data file, with FS scheme. This must match the file_path of the target data file in a manifest entry."
```
Comment (rdblue) [inline site/docs/spec.md:351]: I don't think that requirements for these fields are obvious enough in docs. Let's use the doc field to describe what it is, but the spec requirements should be pulled out and clear.

File: site/docs/spec.md (hunk 2 above)
Comment (chenjunjiedada) [reply site/docs/spec.md]: Just updated. Plus I use the field ID 600 and 601.

File: site/docs/spec.md (hunk 8 above)
Comment (chenjunjiedada) [reply site/docs/spec.md]: OK

File: site/docs/spec.md [hunk 10]
```diff
@@ -333,6 +333,25 @@ Table metadata is stored as JSON. Each table metadata change creates a new table
 
 The atomic operation used to commit metadata depends on how tables are tracked and is not standardized by this spec. See the sections below for examples.
 
+### Delete Format
+
+This section details how to encode row-level deletes in Iceberg metadata. Row-level deletes are not supported in the current format version, 1. This part of the spec is not yet complete and will be completed as format version 2.
```
Comment (leoluan2009) [inline site/docs/spec.md:338]: Can you remove the comma between version and 1?

Comment (chenjunjiedada) [reply site/docs/spec.md]: Sure. I force pushed the commit.

File: site/docs/spec.md [hunk 11]
```diff
@@ -333,6 +333,25 @@ Table metadata is stored as JSON. Each table metadata change creates a new table
...
+
+This section details how to encode row-level deletes in Iceberg metadata. Row-level deletes are not supported in the current format version 1. This part of the spec is not yet complete and will be completed as format version 2.
+
+#### Position-based Delete Files
+
+Position-based delete files identify rows in one or more data files that have been deleted. It has the schema named "position_based_delete_file" with a struct of the following fields:
+
+| Field id, name          | Type                            | Description                                                                                                              |
+|-------------------------|---------------------------------|--------------------------------------------------------------------------------------------------------------------------|
+| **`600 file_path`**     | `required string`               | The full URI of a data file with FS scheme. This must match the file_path of the target data file in a manifest entry.   |
```
Comment (rdblue) [inline site/docs/spec.md:346]: Can you add back-ticks to `file_path` to format it correctly?

File: site/docs/spec.md [hunk 12]
```diff
@@ -333,6 +333,25 @@ Table metadata is stored as JSON. Each table metadata change creates a new table
...
+| **`601 position`**      | `required long`                 | The ordinal position of a deleted row in the target data file identified by file_path, starting at 0.                    |
```
Comment (rdblue) [inline site/docs/spec.md:347]: Here as well.

File: site/docs/spec.md [hunk 13]
```diff
@@ -333,6 +333,25 @@ Table metadata is stored as JSON. Each table metadata change creates a new table
 
 The atomic operation used to commit metadata depends on how tables are tracked and is not standardized by this spec. See the sections below for examples.
 
+### Delete Format
+
+This section details how to encode row-level deletes in Iceberg metadata. Row-level deletes are not supported in the current format version 1. This part of the spec is not yet complete and will be completed as format version 2.
+
+#### Position-based Delete Files
+
+Position-based delete files identify rows in one or more data files that have been deleted. It has the schema named "position_based_delete_file" with a struct of the following fields:
```
Comment (rdblue) [inline site/docs/spec.md:342]: Let's rephrase the second sentence so that it matches the ones above:

> Position-based delete files store `file_position_delete`, a struct with the following fields:

**Comment by rdblue** [comment]
Thanks, @chenjunjiedada! I made a couple of minor changes and merged this.

**Comment by rdblue** [comment]
One of the changes I made was to reset the field IDs in the file/position delete struct. These are separate files, so we we can reuse IDs. The reason why we assign IDs in blocks for the other metadata files is that those are stored in the same manifest file.
